wire-ios-data-model/Source/MLS/CoreCryptoProvider.swift

```diff
@@ -49,30 +58,91 @@ public class CoreCryptoProvider: CoreCryptoProviderProtocol {
         self.cryptoboxMigrationManager = cryptoboxMigrationManager
     }
 
-    // NOTE: this will turn async when we upgrade CC
-    public func coreCrypto(requireMLS: Bool = false) throws -> SafeCoreCryptoProtocol {
-        // TODO: this lock should go away when this function turn async and the class can become an actor
-        return try lock.withLock {
-            let coreCrypto = if let coreCrypto = coreCrypto {
-                coreCrypto
-            } else {
-                try createCoreCrypto()
+    public func coreCrypto(requireMLS: Bool = false) async throws -> SafeCoreCryptoProtocol {
+        let coreCrypto = try await getCoreCrypto()
+
+        if requireMLS {
+            try await initialiseMLS(coreCrypto: coreCrypto)
+        }
+
+        return coreCrypto
+    }
+
+    // Initialise MLS with guranteees that only one task is performing
+    // the operation while others wait for it to complete.
+    //
+    // Based on the structured caching in an actor:
+    // https://forums.swift.org/t/structured-caching-in-an-actor/65501/13
+    private func initialiseMLS(coreCrypto: SafeCoreCrypto) async throws {
+        guard !initialisatingMLS else {
```

Comment:
Do we even need this protection if getting core crypto is already protected and the initialization should only be called once?

Reply:
As discussed offline, yes we need it because initialising MLS is optional, like this:

What if in three separate tasks we execute this:

```
Task1 {
   provider.coreCrypto(requireMLS: false)
}
Task2 {
   provider.coreCrypto(requireMLS: true)
}
Task3 {
   provider.coreCrypto(requireMLS: true)
}
```

Let' s say task 1 runs first and task 2 & 3 are then waiting on the initialising task. When task 1 finishes both task 2 & 3 will continue and both will try to initialise MLS unless we have reentrancy protection.

Reply:
But as pointed out by you I'll add protection against initialising  MLS twice.

Reply:
@caldrian updated